core/string/src/lib.rs

```diff
@@ -1144,3 +1145,348 @@ impl ToStringEscaped for [u16] {
+    /// Caller should ensure that the inner is allocated.
+    #[must_use]
+    unsafe fn data(&self) -> *mut T {
+        // SAFETY:
+        // Caller should ensure that the inner is allocated.
+        unsafe { addr_of_mut!((*self.inner.as_ptr()).data).cast() }
+    }
+
+    #[allow(clippy::cast_ptr_alignment)]
+    fn reserve(&mut self, new_layout: Layout) {
+        // SAFETY:
+        // The layout size of `RawJsString` is never zero, since it has to store
+        // the length of the string and the reference count.
+        // if `realloc` returns valid pointer the old pointer will have been freed
+        // and the data will have been copied to the new memory otherwise the old pointer is still valid.
+        let new_ptr = unsafe { alloc(new_layout) };
+        let Some(new_ptr) = NonNull::new(new_ptr.cast::<RawJsString>()) else {
+            std::alloc::handle_alloc_error(new_layout)
+        };
+
+        if !self.is_dangling() {
+            let old_layout = self.current_layout();
+
+            let old_ptr = self.inner.as_ptr();
+
+            // SAFETY:
+            // New pointer and old pointer are both valid.
+            unsafe {
+                ptr::copy_nonoverlapping(
+                    old_ptr.cast::<u8>(),
+                    new_ptr.as_ptr().cast::<u8>(),
+                    self.initialized_bits(),
+                );
+            }
+            // SAFETY:
+            // Old pointer is valid.
+            unsafe {
+                dealloc(old_ptr.cast(), old_layout);
+            }
+        }
+        self.inner = new_ptr;
+        let new_arr_size = new_layout.size() - DATA_OFFSET;
+        self.cap = new_arr_size / Self::DATA_SIZE;
+    }
+
+    /// Appends an element to the inner `RawJsString` of `JsStringBuilder`.
+    pub fn push(&mut self, v: T) {
+        let len = self.len();
+        if len == self.capacity() {
+            self.expand(len + 1);
+        }
+        // SAFETY:
+        // Capacity has been expanded to be large enough to hold elements.
+        unsafe {
+            self.push_unchecked(v);
+        }
+    }
+
+    /// Push elements from slice to `JsStringBuilder` without doing capacity check.\
+    /// Unlike the standard vector, our holded element types are only `u8` and `u16`, which is [`Copy`] derived,\
+    /// so we only need to copy them instead of cloning.
+    /// # Safety
+    ///
+    /// Caller should ensure the capacity is large enough to hold elements.
+    pub unsafe fn extend_from_slice_unchecked(&mut self, v: &[T]) {
+        // SAFETY: Caller should ensure the capacity is large enough to hold elements.
+        unsafe {
+            ptr::copy_nonoverlapping(v.as_ptr(), self.data().add(self.len()), v.len());
+        }
+        self.len += v.len();
+    }
+
+    /// push elements from slice to `JsStringBuilder`.
+    pub fn extend_from_slice(&mut self, v: &[T]) {
+        let required_cap = self.len() + v.len();
+        if required_cap >= self.capacity() {
```

Comment:
```suggestion
        if required_cap > self.capacity() {
```

Shouldn't it be greater than if the required cap is equal to the current cap then we can copy, right?

Reply:
I tested this before and found sometimes greater and equal condition gives better performance, but maybe this is just a warm-up thing. I think it's fine to just use greater condition.